src/ir/local-graph.h

```diff
@@ -228,7 +232,39 @@ struct LazyLocalGraph : public LocalGraphBase {
     return *locations;
   }
 
+  // Query whether it is valid to move a LocalSet to a new position, that is,
+  // that moving it will not alter observable behavior. That means that no
+  // Localget is able to observe a different value than before. This returns the
+  // set of LocalGets that may do so, that is, the gets for which we detected a
+  // problem, hence if the set is empty, the set is valid to move.
+  //
+  // For example:
+  //
+  //  1. set
+  //  2. get
+  //  3. call
+  //  4. get
+  //
+  // If we move the set to the call, then the get on line 2 could observe a
+  // different value (the pre-existing value in that local, before line 1), and
+  // we'd return that get here.
+  //
+  // |to| must be of the class |queryClass| that is defined during construction.
+  // That is, we must decide ahead of time which places we want to query about
+  // moving LocalSets to, and can then issue queries on any of those, in an
+  // efficient manner.
+  //
+  // This assumes that |to| is in a position dominated by |set|, that is we are
```

Comment:
Not "dominated," just "reached by," right? The whole point is that there might be control flow that skips the destination location, making the move invalid and also meaning that the destination is not dominated by the source.

Reply:
This is the situation we have:
```c
x = C; // set
if (..) goto B;
foo(x); // use
```
We might branch away before reaching the use of `x`, but the use is still dominated by the set (since all paths from the entry to `foo` must go through the set).

Just being reached is not enough:
```c
while (1) {
  foo(x);
  bar(x);
  x = C; // the original set
  if (..) goto B;
}
```
Here the original set reaches `foo(x)` (using the backedge of the loop), but it does not dominate it. Note how `foo(x)`'s influence, were it a set, is larger than `x = C` (`foo` definitely reaches `bar`, while `x = C` might not).

Reply:
Right, I got confused again. Thanks!

Reply:
This is tricky stuff... I keep needing to re-read the sample code.